crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -907,6 +950,84 @@ where
         Ok(())
     }
 
+    /// Invoked if we successfully downloaded a new block from the network.
+    ///
+    /// This will attempt to insert the block into the tree.
+    ///
+    /// There are several scenarios:
+    ///
+    /// ## [BlockStatus::Valid]
+    ///
+    /// The block is connected to the current canonical head and is valid.
+    /// If the engine is still SYNCING, then we can try again to make the chain canonical.
+    ///
+    /// ## [BlockStatus::Accepted]
+    ///
+    /// All ancestors are known, but the block is not connected to the current canonical _head_. If
+    /// the block is an ancestor of the current forkchoice head, then we can try again to make the
+    /// chain canonical, which would trigger a reorg in this case since the new head is therefore
+    /// not connected to the current head.
+    ///
+    /// ## [BlockStatus::Disconnected]
+    ///
+    /// The block is not connected to the canonical head, and we need to download the missing parent
+    /// first.
+    ///
+    /// ## Insert Error
+    ///
+    /// If the insertion into the tree failed, then the block was well-formed (valid hash) but its
+    /// chain is invalid, which means the FCU that triggered the download is invalid. Here we can
+    /// stop because there's nothing to do here and the engine needs to wait for another FCU.
+    fn on_downloaded_block(&mut self, block: SealedBlock) {
+        trace!(target: "consensus::engine", hash=?block.hash, number=%block.number, "Downloaded full block");
+        match self.blockchain.insert_block_without_senders(block) {
+            Ok(status) => {
+                match status {
+                    BlockStatus::Valid => {
+                        // block is connected to the current canonical head and is valid.
+                        self.try_make_sync_target_canonical();
+                    }
+                    BlockStatus::Accepted => {
+                        // block is connected to the canonical chain, but not the current head
+                        self.try_make_sync_target_canonical();
+                    }
+                    BlockStatus::Disconnected { missing_parent } => {
+                        // continue downloading the missing parent
+                        self.sync.download_full_block(missing_parent.hash);
+                    }
```

Comment:
in `try_insert_new_payload` we had this:
```rust
// check if the block's parent is already marked as invalid
if let Some(status) =
    self.check_invalid_ancestor_with_head(block.parent_hash, block.hash)
{
    return Ok(status)
}
```
and `try_insert_new_payload` would be called when handling `EngineSyncEvent::FetchedFullBlock(block)` in `on_sync_event`.
So we should probably add back some logic if the block is disconnected and its parent is invalid?

Reply:
nice find!

Reply:
added